src/data_sources/model/DataVariableListenerManager.ts

```diff
@@ -0,0 +1,62 @@
+import { DataSourcesEvents, DataVariableListener } from '../../data_sources/types';
+import { stringToPath } from '../../utils/mixins';
+import { Model } from '../../common';
+import EditorModel from '../../editor/model/Editor';
+import DataVariable from './DataVariable';
+import ComponentView from '../../dom_components/view/ComponentView';
+import ComponentDataVariable from './ComponentDataVariable';
+
+export interface DataVariableListenerManagerOptions {
+  model: Model | ComponentView;
+  em: EditorModel;
+  dataVariable: DataVariable | ComponentDataVariable;
+  updateValueFromDataVariable: (value: any) => void;
+}
+
+export default class DataVariableListenerManager {
+  private dataListeners: DataVariableListener[] = [];
+  private em: EditorModel;
+  private model: Model | ComponentView;
+  private dataVariable: DataVariable | ComponentDataVariable;
+  private updateValueFromDataVariable: (value: any) => void;
+
+  constructor(options: DataVariableListenerManagerOptions) {
+    this.em = options.em;
+    this.model = options.model;
+    this.dataVariable = options.dataVariable;
+    this.updateValueFromDataVariable = options.updateValueFromDataVariable;
+
+    this.listenToDataVariable();
+  }
+
+  listenToDataVariable() {
+    const { em, dataVariable, model, updateValueFromDataVariable } = this;
+    const { path } = dataVariable.attributes;
+    const normPath = stringToPath(path || '').join('.');
+    const prevListeners = this.dataListeners || [];
+    const [ds, dr] = this.em.DataSources.fromPath(path);
+
+    prevListeners.forEach((ls) => model.stopListening(ls.obj, ls.event, updateValueFromDataVariable));
+
+    const dataListeners: DataVariableListener[] = [];
+    ds && dataListeners.push({ obj: ds.records, event: 'add remove reset' });
+    dr && dataListeners.push({ obj: dr, event: 'change' });
+    dataListeners.push({ obj: dataVariable, event: 'change:value' });
+    dataListeners.push({ obj: em, event: `${DataSourcesEvents.path}:${normPath}` });
```

Comment:
this is the same as the one below

Reply:
https://github.com/GrapesJS/grapesjs/pull/6098